From 36657ddbf213e9ef3f2bda77d4cdbbaaf84c1a3d Mon Sep 17 00:00:00 2001
From: Rui Wang <ruwang@redhat.com>
Date: Fri, 22 Aug 2014 03:52:06 -0400
Subject: [acpi] trace/emca: Add a knob to adjust where to save event log

Message-id: <1408679528-32070-9-git-send-email-ruwang@redhat.com>
Patchwork-id: 88695
O-Subject: [PATCH RHEL7.1 BZ1085141 08/10] trace, eMCA: Add a knob to adjust where to save event log
Bugzilla: 1085141
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

From: "Chen, Gong" <gong.chen@linux.intel.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1085141

Backport of:
Original commit: d6cae935ec5b7873a8ccd8f0331bef2df729e86a
Original Author: "Chen, Gong" <gong.chen@linux.intel.com>
Original Date: Wed, 11 Jun 2014 04:34:50 -0400

 To avoid saving two copies for one H/W event, add a new
 file under debugfs to control how to save event log.
 Once this file is opened, the perf/trace will be used,
 in the meanwhile, kernel will stop printing event log
 to the console. On the other hand, if this file is closed,
 kernel will print event log to the console again.

Signed-off-by: Chen, Gong <gong.chen@linux.intel.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/acpi/acpi_extlog.c b/drivers/acpi/acpi_extlog.c
index 1a32d2c..ff9e9c8 100644
--- a/drivers/acpi/acpi_extlog.c
+++ b/drivers/acpi/acpi_extlog.c
@@ -12,6 +12,7 @@
 #include <linux/cper.h>
 #include <linux/ratelimit.h>
 #include <linux/edac.h>
+#include <linux/ras.h>
 #include <asm/cpu.h>
 #include <asm/mce.h>
 
@@ -156,7 +157,11 @@ static int extlog_print(struct notifier_block *nb, unsigned long val,
  estatus->block_status = 0;
 
  tmp = (struct acpi_generic_status *)elog_buf;
- print_extlog_rcd(NULL, tmp, cpu);
+
+ if (!ras_userspace_consumers()) {
+  print_extlog_rcd(NULL, tmp, cpu);
+  goto out;
+ }
 
  /* log event via trace */
  err_seq++;
@@ -173,6 +178,7 @@ static int extlog_print(struct notifier_block *nb, unsigned long val,
             (u8)gdata->error_severity);
  }
 
+out:
  return NOTIFY_STOP;
 }
 
-- 
1.7.1